src/dune_engine/scheduler.ml

```diff
@@ -1134,6 +1127,23 @@ let flush_file_watcher t =
     Dune_file_watcher.Sync_id.Table.set t.fs_syncs id ivar;
     Fiber.Ivar.read ivar
 
+let wait_for_build_input_change t =
+  match t.wait_for_build_input_change with
+  | Some ivar -> Fiber.Ivar.read ivar
+  | None -> (
+    match t.status with
+    | Standing_by { invalidation; saw_insignificant_changes }
+      when (not (Memo.Invalidation.is_empty invalidation))
+           || saw_insignificant_changes ->
+      Fiber.return ()
+    | Restarting_build _ -> Fiber.return ()
+    | Standing_by _
+    | Building
+    | Shutting_down ->
+      let ivar = Fiber.Ivar.create () in
```

Comment:
If we're shutting down, shouldn't this ivar never return?

Reply:
We could explicitly return `Fiber.never` here, but we don't do it in other functions such as `flush_file_watcher` or `wait_for_process`. It doesn't matter too much what we do here: `shutdown` sends a `Shutdown` event which will blow up the scheduler the next time we reach `Run_once.iter`.

Reply:
I think it would be clearer if it was `never` in those places where it makes sense.

Reply:
That seems fine to me